### PR TITLE
DB:Open should fail on tmpfs when use_direct_reads=true

### PR DESCRIPTION
**Pull request by miasantreble**
Before: 

> $ TEST_TMPDIR=/dev/shm ./db_bench -use_direct_reads=true -benchmarks=readrandomwriterandom -num=10000000 -reads=100000 -write_buffer_size=1048576 -target_file_size_base=1048576 -max_bytes_for_level_base=4194304 -max_background_jobs=12 -readwritepercent=50 -key_size=16 -value_size=48 -threads=32
DB path: [/dev/shm/dbbench]
put error: IO error: While open a file for random read: /dev/shm/dbbench/000007.sst: Invalid argument
put error: IO error: While open a file for random read: /dev/shm/dbbench/000007.sst: Invalid argument
put error: IO error: While open a file for random read: /dev/shm/dbbench/000007.sst: Invalid argument
put error: IO error: While open a file for random read: /dev/shm/dbbench/000007.sst: Invalid argument
put error: IO error: While open a file for random read: /dev/shm/dbbench/000007.sst: Invalid argument
put error: IO error: While open a file for random read: /dev/shm/dbbench/000007.sst: Invalid argument
put error: IO error: While open a file for random read: /dev/shm/dbbench/000007.sst: Invalid argument
put error: IO error: While open a file for random read: /dev/shm/dbbench/000007.sst: Invalid argument
put error: IO error: While open a file for random read: /dev/shm/dbbench/000007.sst: Invalid argument
db_bench: tpp.c:84: __pthread_tpp_change_priority: Assertion `new_prio == -1 || (new_prio >= fifo_min_prio && new_prio <= fifo_max_prio)' failed.
put error: IO error: While open a file for random read: /dev/shm/dbbench/000007.sst: Invalid argument
put error: IO error: While open a file for random read: /dev/shm/dbbench/000007.sst: Invalid argument

After: 
> TEST_TMPDIR=/dev/shm ./db_bench -use_direct_reads=true -benchmarks=readrandomwriterandom -num=10000000 -reads=100000 -write_buffer_size=1048576 -target_file_size_base=1048576 -max_bytes_for_level_base=4194304 -max_background_jobs=12 -readwritepercent=50 -key_size=16 -value_size=48 -threads=32
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
open error: Not implemented: Direct I/O is not supported by the specified DB.

### CODE REVIEW

**Comment by ajkr** [comment]
Well, the drawback is we have to maintain it, and like you said it's hard to make it work for all cases. Isn't it easier to open the IDENTITY file with NewSequentialFile and see if it fails? Not only would it cover direct I/O support on any filesystem, it'd also find any other `env_options_` that are incompatible with the filesystem.

File: db/db_impl_open.cc [hunk 1]
```diff
@@ -356,11 +356,15 @@ Status DBImpl::Recover(
...
     }
-    // Check file system type
-    bool directio_supported = env_->SupportsDirectIO(IdentityFileName(dbname_));
-    if (!directio_supported && immutable_db_options_.use_direct_reads) {
-      return Status::NotSupported(
-          "Direct I/O is not supported by the specified DB.");
+    // Verify compatibility of env_options_ and filesystem
+    {
+      unique_ptr<SequentialFile> idfile;
+      s = env_->NewSequentialFile(IdentityFileName(dbname_), &idfile,
```
Comment (siying) [inline db/db_impl_open.cc]: Although SequentialFile should work, I prefer RandomAccessFile:)

File: db/db_impl_open.cc [hunk 2]
```diff
@@ -356,6 +356,16 @@ Status DBImpl::Recover(
...
       return s;
     }
+    // Verify compatibility of env_options_ and filesystem
+    {
+      unique_ptr<SequentialFile> idfile;
+      s = env_->NewSequentialFile(IdentityFileName(dbname_), &idfile,
+                                  env_options_);
+      if (!s.ok()) {
+        return Status::NotSupported(
+            "Found options incompatible with filesystem. Check your options.");
```
Comment (siying) [inline db/db_impl_open.cc]: Need to pass the information of s into this error message.
Also Status::NotSupported() means RocksDB doesn't support it. If users set up something wrong by themselves, Status::InvalidArgument() is more appropriate.

Comment (miasantreble) [reply db/db_impl_open.cc]: Not sure what's the best way to help use debug the incorrect argument. With `s.getState()` fix, the current message is:
> open error: Invalid argument: Found options incompatible with filesystem: While open a file for random read: /dev/shm/dbbench/IDENTITY: Invalid argument
It's not very clear that -use_direct_reads is the problem here but I guess we can live with it?

Comment (siying) [reply db/db_impl_open.cc]: @miasantreble we don't know whether it's the problem of -use_direct_reads either, right?

File: db/db_impl_open.cc [hunk 3]
```diff
@@ -358,12 +358,12 @@ Status DBImpl::Recover(
...
-      unique_ptr<SequentialFile> idfile;
-      s = env_->NewSequentialFile(IdentityFileName(dbname_), &idfile,
+      unique_ptr<RandomAccessFile> idfile;
+      s = env_->NewRandomAccessFile(IdentityFileName(dbname_), &idfile,
                                   env_options_);
       if (!s.ok()) {
-        return Status::NotSupported(
-            "Found options incompatible with filesystem. Check your options.");
+        return Status::InvalidArgument(
+            "Found options incompatible with filesystem", s.getState());
```
Comment (siying) [inline db/db_impl_open.cc]: I would just put the whole s.ToString() rather than s.getState().

File: db/db_impl_open.cc (hunk 2 above)
Comment (siying) [reply db/db_impl_open.cc]: Another issue: env_options_ doesn't cover all the direct I/O case. We have two options: https://github.com/facebook/rocksdb/blob/master/include/rocksdb/options.h#L584-L601
and test test only covers only one of them. We should set the direct I/O options in a customized EnvOptions when either of the two is true.

**Comment by miasantreble** [comment]
> TEST_TMPDIR=/dev/shm ./db_bench -use_direct_reads=true -benchmarks=readrandomwriterandom -num=10000000 -reads=100000 -write_buffer_size=1048576 -target_file_size_base=1048576 -max_bytes_for_level_base=4194304 -max_background_jobs=12 -readwritepercent=50 -key_size=16 -value_size=48 -threads=32
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
open error: Not implemented: Direct I/O is not supported by the specified DB.

File: db/db_impl_open.cc [hunk 4]
```diff
@@ -356,6 +356,26 @@ Status DBImpl::Recover(
       assert(s.IsIOError());
       return s;
     }
+    // Verify compatibility of env_options_ and filesystem
+    {
+      unique_ptr<RandomAccessFile> idfile;
+      s = env_->NewRandomAccessFile(IdentityFileName(dbname_), &idfile,
+                                  env_options_);
+      if (!s.ok()) {
+        const char* errorMsg = s.ToString().c_str();
```
Comment (ajkr) [inline db/db_impl_open.cc]: the convention is lower-case underscore-separated variable names

File: db/db_impl_open.cc [hunk 5]
```diff
@@ -356,6 +356,26 @@ Status DBImpl::Recover(
...
+        // Check if use_direct_reads is the root cause
+        s = env_->NewRandomAccessFile(IdentityFileName(dbname_), &idfile,
+                                    env_->OptimizeForLogRead(env_options_));
+        if (s.ok()) {
+          return Status::NotSupported(
+              "Direct I/O is not supported by the specified DB.");
+        }
+        else {
```
Comment (ajkr) [inline db/db_impl_open.cc]: no line break. maybe `make format` would catch it?

**Comment by miasantreble** [comment]
Log with commit c72da44

> TEST_TMPDIR=/dev/shm ./db_bench -use_direct_io_for_flush_and_compaction=true -benchmarks=readrandomwriterandom -num=10000000 -reads=100000 -write_buffer_size=1048576 -target_file_size_base=1048576 -max_bytes_for_level_base=4194304 -max_background_jobs=12 -readwritepercent=50 -key_size=16 -value_size=48 -threads=32
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
open error: Not implemented: Direct I/O is not supported by the specified DB.

> TEST_TMPDIR=/dev/shm ./db_bench -use_direct_reads=true -benchmarks=readrandomwriterandom -num=10000000 -reads=100000 -write_buffer_size=1048576 -target_file_size_base=1048576 -max_bytes_for_level_base=4194304 -max_background_jobs=12 -readwritepercent=50 -key_size=16 -value_size=48 -threads=32
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
open error: Not implemented: Direct I/O is not supported by the specified DB.

**Pull request closed without merging**